rsocket-micrometer/src/main/java/io/rsocket/micrometer/MicrometerRSocket.java

```diff
@@ -0,0 +1,208 @@
+    Objects.requireNonNull(meterRegistry, "meterRegistry must not be null");
+
+    this.metadataPush = new InteractionCounters(meterRegistry, "metadata.push", tags);
+    this.requestChannel = new InteractionCounters(meterRegistry, "request.channel", tags);
+    this.requestFireAndForget = new InteractionCounters(meterRegistry, "request.fnf", tags);
+    this.requestResponse = new InteractionTimers(meterRegistry, "request.response", tags);
+    this.requestStream = new InteractionCounters(meterRegistry, "request.stream", tags);
+  }
+
+  @Override
+  public void dispose() {
+    delegate.dispose();
+  }
+
+  @Override
+  public Mono<Void> fireAndForget(Payload payload) {
+    return delegate.fireAndForget(payload).doFinally(requestFireAndForget);
+  }
+
+  @Override
+  public Mono<Void> metadataPush(Payload payload) {
+    return delegate.metadataPush(payload).doFinally(metadataPush);
+  }
+
+  @Override
+  public Mono<Void> onClose() {
+    return delegate.onClose();
+  }
+
+  @Override
+  public Flux<Payload> requestChannel(Publisher<Payload> payloads) {
+    return delegate.requestChannel(payloads).doFinally(requestChannel);
+  }
+
+  @Override
+  public Mono<Payload> requestResponse(Payload payload) {
+    return Mono.defer(
+        () -> {
+          Sample sample = requestResponse.start();
+
+          return delegate
+              .requestResponse(payload)
+              .doFinally(signalType -> requestResponse.accept(sample, signalType));
+        });
+  }
+
+  @Override
+  public Flux<Payload> requestStream(Payload payload) {
+    return delegate.requestStream(payload).doFinally(requestStream);
+  }
+
+  private static final class InteractionCounters implements Consumer<SignalType> {
+
+    private final Counter cancel;
+
+    private final Counter onComplete;
+
+    private final Counter onError;
+
+    private InteractionCounters(MeterRegistry meterRegistry, String interactionModel, Tag... tags) {
+      this.cancel = counter(meterRegistry, interactionModel, CANCEL, tags);
+      this.onComplete = counter(meterRegistry, interactionModel, ON_COMPLETE, tags);
+      this.onError = counter(meterRegistry, interactionModel, ON_ERROR, tags);
+    }
+
+    @Override
+    public void accept(SignalType signalType) {
+      switch (signalType) {
+        case CANCEL:
+          cancel.increment();
+          break;
+        case ON_COMPLETE:
+          onComplete.increment();
+          break;
+        case ON_ERROR:
+          onError.increment();
+          break;
+      }
+    }
+
+    private static Counter counter(
+        MeterRegistry meterRegistry, String interactionModel, SignalType signalType, Tag... tags) {
+
+      return meterRegistry.counter(
+          String.format("rsocket.%s", interactionModel),
+          Tags.of(tags).and("signal.type", signalType.name()));
+    }
+  }
+
+  private static final class InteractionTimers implements BiConsumer<Sample, SignalType> {
+
```

Comment:
One thing I was thinking about adding in something else I was working on was the idea of capturing the time between onNexts. Right now it's really hard to tell how quickly, or slowly a stream is emitting data. Having the time between when the stream starts and cancel/error/complete doesn't really tell that.

Reply:
Nice - is that a https://micrometer.io/docs/concepts#_distribution_summaries

Can we iterate on improvements like this? I'd love to get this in sooner and start seeing what results look like from the CLI etc.

Reply:
Yep - something like that.

`Can we iterate on improvements like this? I'd love to get this in sooner and start seeing what results look like from the CLI etc.`
Sure

Reply:
Agreed that's a really useful metric, but also that we should kick the can on it.  At some level we're getting rates at the `DuplexConnection`, but I'll confer with @jkschneider and work on #487 once I have a plan.

Reply:
@robertroeser I think this `onNext` timer would be potentially an interesting application of the `IntervalTimer` we plan to introduce with Micrometer 1.1. It's purpose is to count/time the space between things. Conventionally I've seen it used as something to alert me when say a periodic batch job doesn't kick off in an expected amount of time, but the concept interestingly applies equally well here.